File: examples/cpp/sniff.cpp
#include <csignal>
#include <cstdio>
#include <iostream>
#include <stdexcept>

#include "netforge/io/sniffer.hpp"
#include "netforge/layers/dns.hpp"
#include "netforge/layers/ethernet.hpp"
#include "netforge/layers/icmp.hpp"
#include "netforge/layers/ipv4.hpp"
#include "netforge/layers/ipv6.hpp"
#include "netforge/layers/tcp.hpp"
#include "netforge/layers/udp.hpp"

using namespace netforge;

static volatile sig_atomic_t g_stop = 0;

static void on_signal(int) { g_stop = 1; }

int main(int argc, char* argv[]) {
    const char* dev = argc > 1 ? argv[1] : "lo";
    const char* filter = argc > 2 ? argv[2] : "";

    std::signal(SIGINT, on_signal);

    std::printf("interfaces:\n");
    for (const auto& name : list_interfaces()) std::printf("  %s\n", name.c_str());

    sniffer s(dev, filter);
    std::printf("capturing on %s with filter '%s' (Ctrl-C to stop)\n", dev, filter);

    uint8_t* data = nullptr;
    std::size_t len = 0;
    while (!g_stop) {
        if (!s.next(data, len)) return 1;
        std::printf("%4zu bytes  ", len);

        buffer in(data, len);
        try {
            ethernet_header_t eth = ethernet_header_t::consume(in);
            std::cout << eth;
            if (eth.ethertype == 0x0800) {
                ipv4_header_t ip = ipv4_header_t::consume(in);
                std::cout << "  " << ip;
                if (ip.protocol == 17) {
                    std::cout << "  " << udp_header_t::consume(in);
                } else if (ip.protocol == 6) {
                    std::cout << "  " << tcp_header_t::consume(in);
                } else if (ip.protocol == 1) {
                    std::cout << "  " << icmp_header_t::consume(in);
                }
            } else if (eth.ethertype == 0x86DD) {
                std::cout << "  " << ipv6_header_t::consume(in);
            }
        } catch (const std::out_of_range& ex) {
            std::printf("  (truncated frame: %s)", ex.what());
        }
        std::printf("\n");
    }
    return 0;
}

File: examples/python/sniff.py
import signal
import sys

import netforge

filter_expr = sys.argv[1] if len(sys.argv) > 1 else "udp port 53"
dev = sys.argv[2] if len(sys.argv) > 2 else "lo"

print("interfaces:", ", ".join(netforge.list_interfaces()))
sniffer = netforge.Sniffer(dev, filter_expr)
print(f"capturing on {dev} with filter '{filter_expr}' (Ctrl-C to stop)")

signal.signal(signal.SIGINT, lambda *_: sys.exit(0))

while True:
    raw = sniffer.next()
    if raw is None:
        continue
    eth = netforge.EthernetHeader.from_bytes(raw)
    print(f"{len(raw):4} bytes  eth 0x{eth.ethertype:04x}")
    if eth.ethertype == 0x0800:
        ip = netforge.Ipv4Header.from_bytes(raw[14:34])
        print(f"           ipv4 ttl={ip.ttl} proto={ip.protocol} "
              f"{netforge.ipv4_to_string(ip.source_ip)} -> "
              f"{netforge.ipv4_to_string(ip.destination_ip)}")
        if ip.protocol == 17 and len(raw) >= 42:
            udp = netforge.UdpHeader.from_bytes(raw[34:42])
            print(f"           udp {udp.src_port} -> {udp.dst_port}")
            if udp.dst_port == 53 or udp.src_port == 53:
                dns = netforge.DnsMessage.from_bytes(raw[42:])
                for q in dns.questions:
                    print(f"           dns q: {q.qname}")